### ISSUE TITLE
proposal: 2404 indexedDB -> inactive

### ISSUE THREAD

**Issue by Thorin-Oakenpants** [ghacksuserjs/ghacks-user.js#226]
IndexedDB is problematic.

### :tractor: issue 1: internals
It uses your `profile/storage` folder which has three folders, one of which is used internally by FF ( the `permanent` folder). Disabling indexedDB causes errors on Firefox startup for example

![noindexeddb](https://user-images.githubusercontent.com/16656956/29164294-101da666-7e13-11e7-8636-2a64588207e5.png)

Note: I have had indexedDB disabled for 2+ years with no FF problems that I know of up to FF54. Since 55 I have had it enabled as I am especially weary of issues arising from WE and FF migration to 57.

### :blue_car: issue 2: add-ons & web ext
Some legacy and web ext's use indexedDB. While I have not seen any issues with this my add-ons up til now, it looks like it **may** be an issue with Web Exts (I am not a dev, but I believe the methods for local storage under WE are more limited eg no longer using extensions.* prefs)

### :taxi: issue 3: websites
Websites can pack a sad when indexedDB is disabled. EVEN when it is not used. 2+ years and I still do not have a problem with this really. I do not think it is widespread, but I lack any metrics

### :ambulance: issue 4: control <sup>someone better call a wambulance</sup>
There are NO mechanisms currently within FF for dealing with indexedDB. You cannot clear by host or control access per domain (yet!). Because I have never bothered with indexedDB, I am not even sure it can be cleared on close, certainly not with it's own pref (but may be bundled with clearing something else - **someone please let me know**)

This may change (here's hoping) as the move to WE API and requests on these have opened new doors - eg see [1333050](https://bugzilla.mozilla.org/show_bug.cgi?id=1333050) .. which has a sort of meta ticket see [1340511](https://bugzilla.mozilla.org/show_bug.cgi?id=1340511) (yes, its P5 because it is _sooo broad_). Also do not forget the new permissions UI coming, which may help play a part.

### :fire_engine:  notes <sup>yes, that's a fire engine</sup>

- FYI: indexedDB is disabled in PB mode (because there are no mechanisms to control this persistent storage, they had to leave it out)
- indexedDB is covered by FPI. This does not solve the issue of 1st party tracking but does mitigate 3rd party, and even enhances the granular level of uMatrix
   * turn on indexedDB + FPI and visit youtube, you should see `profile/storage/default` entries such as `https+++www.youtube.com^firstPartyDomain=youtube.com`
- a ~~workaround~~ hack is to set restrictive permissions on the `profile/storage/default` and `profile/storage/temporary` directories (see [first post here](https://www.wilderssecurity.com/threads/firefox-and-indexeddb.395113/). This may make some sites work better by tricking them, but does nothing for sites that would still break
- you could use external apps to clean the two above folders

### conclusion
There are three upsides to enabling indexedDB (issues 1,2,3) and a somewhat mitigated (notes) downside (issue 4). I am more concerned that at FF55+ we should flip this for migration purposes to Web Ext - it's a bumpy ride. That however, does not help ESR users.

Class, discuss!

**Comment by 2glops** [comment]
uBo hybrid webext 1.4.0 required indexedDB for assets storage:
https://github.com/gorhill/uBlock/releases/tag/1.14.0

IMO more and more extensions will use indexedDB.
So yes, making it inactive is a good path.

**Comment by Thorin-Oakenpants** [comment]
Well I guess that seals the deal then. uBo is pretty much essential

**Comment by arno29** [comment]
Please wait with the change to inactive. Let's wait for the release of uBlock 1.14.2 in the AMO. Dom.indexdDb some sites use to store cookies and metadata. Firefox does not clear this store after closing. Only external program CCleaner.

**Comment by Thorin-Oakenpants** [comment]
@arno ... why wait for 14.2 when 14.0 is the one that will need indexedDB? The change should also be well in advance, before the releases on AMO. It's a shame the 55 alpha doesn't have this proposed change in it, as I now fear it will cause issues for users not right up to date.

**Comment by arno29** [comment]
I left **dom.indexeddb.true** active (false) and do not see any difference in the work of uBlock 1.14.3b0. It's not wise to leave it "true". Potential vector of attack and violation of privacy.

**Comment by Thorin-Oakenpants** [comment]
I don't have time or means to test this with versions of uBo+FF and what happens to cached assets, sorry. These are betas versions right? And you're using on 57 or something?

_[1] "Cacheable assets" refer to: filter lists downloaded from remote servers; compiled (pre-parsed) filter lists; "selfie" (kind of dated doc but "selfie" part is still relevant)_

indexedDB will **have** to => inactive at some stage, and **soon**. It is going to break too many things. There are other web extensions using indexedDB as well, eg Stylus/ViolentMonkey (one of them at least, the other changed). I guess it's a thing now as compared to legacy add-ons.

Probably the same will happen with cookies - we will probably not be able to block ALL by default as this causes issues for WE re access to local dom storage - but that's another discussion

Someone please confirm that indexeddb => disabled => console errors/performance issues => uBo 1.14+ : i.e that indexedDB needs to be enabled

**Comment by arno29** [comment]
Yes, it`s beta, but on Firefox 55.0.3. dom.indexdDb storage basically use websites that want to know everything about the user. Like google. So, the percentage of broken sites is very small.

**Comment by Thorin-Oakenpants** [comment]
@arno29 Correct. As per the first post. indexedDB is used by websites, with first party isolation this protects from 3rd parties, but does nothing to stop being tracked on subsequent visits to first parties, and there are no decent mechanisms to deal with cleaning indexeddb. These are the downsides.

The percentage of broken sites will vary from user to user. I personally have had indexeddb disabled for years, until now.

The point is, ***there is no point disabling indexedDB if it starts to break extensions and even Firefox itself***, and we'll just have to suck it up and deal with the fallout (eg folder permissions, ccleaner, pb mode) until a proper solution arises

**Comment by arno29** [comment]
I agree. You again face the choice to please one user or another. Such controversial points have already been. It's up to you to decide.

**Comment by Thorin-Oakenpants** [comment]
> do not see any difference in the work of uBlock 1.14.3b0

I want these cleared up
- Is the pref `dom.enable_resource_timing` required to be `true` in order for Web Extensions to use indexedDb for storage?
   * I believe so, since when it is false even Firefox itself throws console errors
- Therefore ... If the pref `dom.enable_resource_timing` is set to `false`, then uBo is not working as designed (but it may work as it is only used, for now, for _cachable assets_ ). IDK. I am not a uBo wizard

**Comment by earthlng** [comment]
https://github.com/gorhill/uBlock/issues/2925

**Comment by crssi** [comment]
^^This might be the problem here (FF 55.0.3 32bit). Will try today with 64bit.
Since when I use many lists in uBO 1.13.10 or 1.13.11rc2 (1.13.8 works fine) then Firefox get choking.
gorhill was pissed at me yesterday, but it was my communication skill or non-skill failure, since I failed to see his instructions posted right at time I have testing... my bad :(, but I never said its uBO fault.
I have really a great respect towards gorhill

**Comment by earthlng** [comment]
I agree with Pants here and more Webextensions will migrate to use IDB.
Hopefully once the required APIs are implemented to deal with IDB, someone will create an awesome WE to selectively delete IDBs, ideally with whitelisting support.
Until then either use PB window for sites that you allow to use JS and are using IDB, or clear the IDB storage manually or with a tool like CCleaner.

**Comment by Thorin-Oakenpants** [comment]
I have an issue, swine I updated to 1.13.10 (on 55, and since updated to 55.0.3), I cannot view standalone images i.e http* (file:/// is fine). The object seems to load, but nothing shows or is drawn, the previous page is still visible. Nothing in the logger. And uMatrix went silly as well (I have NOT touched it), and list updates starting failing (XHR blocked), and XHR was neutral as it has always been. I went back and checked my FF54 settings. Its all breaking. I almost prepared to do something drastic

**Comment by earthlng** [comment]
you can try and delete the startupCache folder. Sometimes that helped me when I had weird problems after an update.

**Comment by Thorin-Oakenpants** [comment]
I just rolled back to 1.13.8, removed the XHR rule from uMatrix as well - and everything is working.

I was just readng about that startupcache but didn;t know who to run the switch on a portable FF. Yeah, so just delete it then, sounds good. I have a full backup as of a minute ago. I'll do another uBo update in a few days

**Comment by arno29** [comment]
uBlock 1.14.3b0
uMatrix 1.0.1rc0
user_pref("dom.indexedDB.enabled", false);

After 9 hours of work, not a single mistake was noticed. Everything is as usual. All filters work. Maybe there is some sort of a way to reproduce the problem?

**Comment by Thorin-Oakenpants** [comment]
indexedDB is a mess - it needs an overhaul - here are two more bugs of interest

[1192643](https://bugzilla.mozilla.org/show_bug.cgi?id=1192643) - window.indexedDB throws when dom.indexedDB.enabled=false
[1323974](https://bugzilla.mozilla.org/show_bug.cgi?id=1323974) - IndexedDB - dom.indexedDB.enabled set to false doesn't work properly

**Comment by earthlng** [comment]
That quote is from a guy who said about himself

> I'm not tech savvy and I don't understand the Firefox systems deeply.

and yet that sure didn't stop him from making statements like

> This is total & pure madness.

> Here we have a good example how Mozilla people doesn't understand AT ALL about end-users privacy!

And then

> It doesn't disable asm.js and other caches Andrew Sutherland mentions in his 1st comment.

... no of course not, why would it?!
There are lots of problems with FF but this is not one of them! And *his* problem has absolutely nothing to do with IndexedDB

**Comment by Thorin-Oakenpants** [comment]
Well, it sounded scary and hasn't been marked invalid. You didn't actually expect me to *read* it did you? lulz

Anyway, I think we can close this

**Comment by Thorin-Oakenpants** [comment]
@earthlng This describes the issue I was having with uBo -> 1.14 ( https://www.reddit.com/r/firefox/comments/6x1yec/address_bar_randomly_doesnt_work_after_recent/dmco0fy/ ). Except for me it was only standalone images, and F5 did nothing. I rolled back. And as for my smartphone (FF55.02, uBo seems to be working but I cannot access the options - this after the update did some weird stuff, and I ended up uninstalling it, force quitting the app, clearing app cache, restarting, reinstalling 1.14). IDK what to do now re smartphone - guess it's uninstall, reinstall FF time.

Am so glad uMatrix is now delayed until 56 :) One thing at a time, son.
